doc/Environments.md

```diff
@@ -0,0 +1,78 @@
+# Environments
+
+We use [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) for any environment related work.
+
+For more documentation on environments, read the references listed at the bottom.
+
+## Runtime environments
+
+By default, the template offers the following runtime environments, defined by their corresponding `appsettings` files.
+
+- Development (`appsettings.development.json`)
+- Staging (`appsettings.staging.json`)
+- Production (`appsettings.production.json`)
+
+You can add / remove runtime environments by simply adding or removing appsettings files (e.g. `appsettings.myenvironment.json` will create a new runtime environment named `myenvironment`).
+
+This is configured inside the [AppSettingsConfiguration.cs](../src/app/ApplicationTemplate.Shared/Configuration/AppSettingsConfiguration.cs) file.
+
+- The default runtime environment is set based on a compile-time directive (e.g. production).
+
+- You can get the current environment using `AppSettingsConfiguration.AppEnvironment.GetCurrent`.
+
+- You can get all the possible environments using `AppSettingsConfiguration.AppEnvironment.GetAll`.
+
+- You can set the current environment using `AppSettingsConfiguration.AppEnvironment.SetCurrent`. If the environment doesn't exist, you will get an exception.
+
+- The current environment is persisted into a file that is processed at startup; this allows the environment to affect your IoC and be faster than accessing any settings service (e.g. no IoC, no deserialization, etc.).
+
+- The current environment is also used when configuring the `IHostBuilder`. For example, you can do the following:
+  ```csharp
+  hostBuilder.ConfigureServices((context, services) => {
+    var isDevelopment = context.HostingEnvironment.IsDevelopment();
+    var isStaging = context.HostingEnvironment.IsStaging();
+    var isProduction = context.HostingEnvironment.IsProduction();
+    var environment = context.HostingEnvironment.EnvironmentName;
+    var isMyEnvironment = context.HostingEnvironment.IsEnvironment("MyEnvironment");
+  });
+  ```
+
+## Compile-time environments
+
+Compile-time environments should be avoided as much as possible. It's always preferable to properly support runtime environments. Compile-time environments require a different compilation of the code which is affected most of the time using compilation directives (e.g. `#if PRODUCTION`) and constants.
```

Comment:
Should we say that the compilation constants should be used only in .csproj, and not in .cs files?

Reply:
Do you think it adds value? We can add that if we think it's not clear!

Reply:
`<DefinedConstants>` should not be avoided entirely, since platform constants (__IOS__, __WASM__) which are part of this group, are really useful or necessary. They are widely used in uno, framework, or in client-code.
But, I agree that constants like __DEBUG__ should be avoided as much as possible.

Reply:
I agree with Xiao, this is really for environment constants. We'll leave it like that. Thanks guys 👍